### PR TITLE
Replace login event handlers

### LINKED ISSUE #286: Replace log in/out event handlers with Redux behaviours
## Description

As a move towards eliminating event listeners in the canoe client side code base there is a desire to try replacing the login and logout event handlers in `app.riot.html` with functionality based on Redux.

## Requirements

- [ ] Replace all event handling code related to user log in and log out behaviours with Redux based functionality.

### PR DESCRIPTION
**Pull request by md8n**
Closes #286 

## Description

To move away from using event handlers to Redux this is a first 'experiment' to use Redux state change to trigger other actions.

This uses the library `redux-pure-subscribe` for watching state changes in the Redux store.

## Instructions

Run `yarn` to refresh the packages.  And then `yarn start` pointing at your favourite `canoe` backend service.

Try logging in, and ... wait for it ... , logging out ...

### CODE REVIEW

**Comment by boerwastaken** [comment]
I yarned but it yearns for redux-pure-subscribe.
```
RiotReduxPlugin.js:1 Uncaught (in promise) Error: Cannot find module 'redux-pure-subscribe'
    at webpackMissingModule (RiotReduxPlugin.js:1)
    at Module../src/js/redux/RiotReduxPlugin.js (RiotReduxPlugin.js:1)
    at __webpack_require__ (bootstrap:68)
    at Module../src/app.js (app.js:1)
    at __webpack_require__ (bootstrap:68)
```

I wanted to run this branch to try and see if your reduxing breaks the clean state app booting as enabled in #274 and #281 .

If you can 1) delete your localStorage and cookies 2) reload 3) click around some lessons, you haven't broken it.

Also, keep in mind that only the server can tell you whether you're currently logged in or not.

File: src/riot/Components/LoadCompletions.riot.html [hunk 1]
```diff
@@ -38,7 +38,8 @@
                 await this.pullIdbDataIntoMemory();
 
                 signalCompletionsAreReady();
-                dispatchLoggedInEvent();
+                // Normally this will have already occured in AuthenticationUtilities.login
+                signalUserLoggedIn();
```
[the remark this answers is not in the document]
Comment (md8n) [reply src/riot/Components/LoadCompletions.riot.html:42]: This is the direct equivalent of what was in this code before.  I haven't dug all of the way through things to understand whether this second flagging of the user having logged in is superfluous or not.

File: src/riot/Components/LoadCompletions.riot.html [hunk 2]
```diff
@@ -47,11 +48,18 @@
             },
 
             async clearAppData() {
+                if (!!this.state.clearingAppData) {
+                    return;
+                }
+                this.state.clearingAppData = true;
```
[the remark this answers is not in the document]
Comment (md8n) [reply src/riot/Components/LoadCompletions.riot.html:54]: Riot doesn't do a great job of identifying that an `update` is already under way (it doesn't do it at all), so this semaphore helps the following bit of code to not be fired multiple times (it triggers an `update` and `update` calls it).

File: src/riot/App.riot.html [hunk 3]
```diff
@@ -130,33 +131,56 @@
...
+                        }
+
+                        this.state.logoutUpdating = true;
+                        logoutUpdate();
+                        this.state.logoutUpdating = false;
+                    }
+                };
+    
+                this.state.storeSubscriptions["completion"] = storeSubscribers(completionCallback, completionTargets);
+                this.state.storeSubscriptions["logout"] = storeSubscribers(logoutCallback, logoutTargets);
```
[the remark this answers is not in the document]
Comment (md8n) [reply src/riot/App.riot.html:176]: This is a more focused approach than the `storeListener` below, and does a different job.

With this approach the callback is only called when any of the specific targets change (as opposed to when anything within the entire store changes such as `storeListener` does).  This means that we can easily replace event handlers by watching specific fields within the store.

**Comment by PeteCoward** [comment]
> If you can 1) delete your localStorage and cookies 2) reload 3) click around some lessons, you haven't broken it.

I did this, it allowed me to click around and view content, then it redirected me to login when it received a 403 from the actions POST api on completing a card
When I turned network off, it allowed me to click around just fine, but the images in RichText blocks were not available offline ( which is expected behaviour now we don't cache rich text images )

**Comment by boerwastaken** [comment]
> I did this, it allowed me to click around and view content [...]
you may have been doing something differently, because for me it didn't work quite right, but neither did it with 9cff9000.


>  I think we agree that how we achieve that needs more thought.

It does. The current implementation doesn't actually work for all cases, because `ReduxImpl/Interface.getManifestFromStore()` doesn't actually always give you the manifest (I was enticed by the name, but was also wondering how it would do that in synchronous code, since the manifest is in a Cache... I shouldn't have been so trusty). While a probe for a manifest is an appropriate method to determine *"is this a from scratch boot, or are we running on teleported caches"*, the answer can only come in an async way, while the code that needs to know — Riot template conditionals — is synchronous. So that needs some rethinking.

**Comment by PeteCoward** [comment]
marking as draft, will be reworked as part of new philosophy

**Pull request closed without merging**